lib/Nodes/TableNode.php

```diff
@@ -4,177 +4,435 @@
+
+        $columns = 0;
+        foreach ($this->data as $row) {
+            $columns = max($columns, count($row->getColumns()));
+        }
+
+        return $columns;
     }
 
     public function getRows() : int
     {
-        return count($this->data) - 1;
+        if ($this->isCompiled === false) {
+            throw new LogicException('Call compile() first.');
+        }
+
+        return count($this->data);
     }
 
     /**
-     * @return string[][]|SpanNode[][]
+     * @return TableRow[]
      */
     public function getData() : array
     {
+        if ($this->isCompiled === false) {
+            throw new LogicException('Call compile() first.');
+        }
+
         return $this->data;
     }
 
     /**
+     * Returns an of array of which rows should be headers,
+     * where the row index is the key of the array and
+     * the value is always true.
+     *
      * @return bool[]
      */
     public function getHeaders() : array
     {
+        if ($this->isCompiled === false) {
+            throw new LogicException('Call compile() first.');
+        }
+
         return $this->headers;
     }
 
-    /**
-     * @param mixed[]|null $parts
-     */
-    public function push(?array $parts, string $line) : bool
+    public function pushSeparatorLine(Parser\TableSeparatorLineConfig $separatorLineConfig) : void
     {
-        $line = utf8_decode($line);
+        if ($this->isCompiled === true) {
+            throw new LogicException('Cannot push data after TableNode is compiled');
+        }
 
-        if ($parts !== null) {
-            // New line in the tab
-            if ($parts[2] !== $this->parts[2]) {
-                return false;
-            }
+        $this->separatorLineConfigs[$this->currentLineNumber] = $separatorLineConfig;
+        $this->currentLineNumber++;
+    }
+
+    public function pushContentLine(string $line) : void
+    {
+        if ($this->isCompiled === true) {
+            throw new LogicException('Cannot push data after TableNode is compiled');
+        }
+
+        $this->rawDataLines[$this->currentLineNumber] = utf8_decode($line);
+        $this->currentLineNumber++;
+    }
+
+    public function finalize(Parser $parser) : void
+    {
+        if ($this->isCompiled === false) {
+            $this->compile();
+        }
 
-            if ($this->type === self::TYPE_PRETTY) {
-                if ($parts[0] === true) {
-                    $this->headers[count($this->data) - 1] = true;
+        $tableAsString = $this->getTableAsString();
+
+        if (count($this->errors) > 0) {
+            $parser->getEnvironment()
+                ->getErrorManager()
+                ->error(sprintf("%s\nin file %s\n\n%s", $this->errors[0], $parser->getFilename(), $tableAsString));
+
+            $this->data = [];
+
+            return;
+        }
+
+        foreach ($this->data as $i => $row) {
+            foreach ($row->getColumns() as $col) {
+                // TODO - not sure if this ever did anything
```

Comment:
Can we do something about this? If not, lets remove it and add an issue for it to address later.

Reply:
I removed it. I literally can't think of how that line was *ever* hit. So, it either never did anything, or it's so edge-case that I can't even figure out *what* issue I'd create. I think we roll forward without it, and see if any issues come up. Clearly we didn't have a test for the functionality it gave us, if any.